File: examples/go/bitwise-ops/BitwiseOps.runar.go
package contract

import runar "github.com/icellan/runar/packages/runar-go"

// BitwiseOps demonstrates bitwise and shift operators on Bigint values.
type BitwiseOps struct {
	runar.SmartContract
	A runar.Bigint `runar:"readonly"`
	B runar.Bigint `runar:"readonly"`
}

// TestShift verifies shift operators compile and run.
func (c *BitwiseOps) TestShift() {
	left := c.A << 2
	right := c.A >> 1
	runar.Assert(left >= 0 || left < 0)
	runar.Assert(right >= 0 || right < 0)
	runar.Assert(true)
}

// TestBitwise verifies bitwise operators compile and run.
func (c *BitwiseOps) TestBitwise() {
	andResult := c.A & c.B
	orResult := c.A | c.B
	xorResult := c.A ^ c.B
	notResult := ^c.A
	runar.Assert(andResult >= 0 || andResult < 0)
	runar.Assert(orResult >= 0 || orResult < 0)
	runar.Assert(xorResult >= 0 || xorResult < 0)
	runar.Assert(notResult >= 0 || notResult < 0)
	runar.Assert(true)
}

File: examples/rust/bitwise-ops/BitwiseOps.runar.rs
use runar::prelude::*;

/// BitwiseOps — Demonstrates bitwise and shift operators on Bigint values.
#[runar::contract]
pub struct BitwiseOps {
    #[readonly]
    pub a: Bigint,
    #[readonly]
    pub b: Bigint,
}

#[runar::methods(BitwiseOps)]
impl BitwiseOps {
    /// Verify shift operators compile and run.
    #[public]
    pub fn test_shift(&self) {
        let left = self.a << 2;
        let right = self.a >> 1;
        assert!(left >= 0 || left < 0);
        assert!(right >= 0 || right < 0);
        assert!(true);
    }

    /// Verify bitwise operators compile and run.
    #[public]
    pub fn test_bitwise(&self) {
        let and_result = self.a & self.b;
        let or_result = self.a | self.b;
        let xor_result = self.a ^ self.b;
        let not_result = ~self.a;
        assert!(and_result >= 0 || and_result < 0);
        assert!(or_result >= 0 || or_result < 0);
        assert!(xor_result >= 0 || xor_result < 0);
        assert!(not_result >= 0 || not_result < 0);
        assert!(true);
    }
}
